🎉Source HubSpot: Adds form_submission and property_history streams

tinomerl opened the pull request:
## What
* Adds Form Submission Stream described in #6493
* Adds Property History Stream described in #7554 

## How
*Describe the solution*

## Recommended reading order
1. `x.java`
2. `y.python`

## Pre-merge Checklist
Expand the relevant checklist and delete the others. 

<details><summary> <strong> New Connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/SUMMARY.md`
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
   
</p>
</details>


<details><summary> <strong> Updating a connector </strong></summary>
<p>
   
#### Community member or Airbyter
   
- [ ] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret` 
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated 
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
   
#### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items. 
   
- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] Credentials added to Github CI. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci). 
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing. 
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the new connector version is published, connector version bumped in the seed directory as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</p>
</details>

<details><summary> <strong> Connector Generator </strong> </summary>
<p>
   
- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.io/contributing-to-airbyte/updating-documentation#issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed.
</p>
</details>

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/airbytehq/airbyte/7787)
<!-- Reviewable:end -->

marcosmarxm commented on airbyte-integrations/connectors/source-hubspot/Dockerfile:
```diff
@@ -34,5 +34,5 @@ COPY source_hubspot ./source_hubspot
 ENTRYPOINT ["python", "/airbyte/integration_code/main.py"]
 
-LABEL io.airbyte.version=0.1.23
+LABEL io.airbyte.version=0.2.0
```
Why changing to 0.2.0 and not 0.1.24?

tinomerl replied:
@marcosmarxm thanks for the feedback. I bumped the minor version up since the two new streams add functionality and are not bugfixes.

marcosmarxm commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:
```diff
@@ -316,7 +316,7 @@ def _field_to_datetime(value: Union[int, str]) -> pendulum.datetime:
         """Skip records that was updated before our start_date"""
         for record in records:
-            updated_at = record[self.updated_at_field]
+            updated_at = record.get(self.updated_at_field) or None
```
the `.get` method already use None as default

marcosmarxm commented:
I tried to run the tests locally  and didn't work. Can you share the results from your tests, check https://docs.airbyte.io/connector-development/testing-connectors/source-acceptance-tests-reference#architecture-of-standard-tests or the checklist **Updating connector**

marcosmarxm commented:
> @marcosmarxm currently running in timeouts when running the acceptance tests for basic_read and full_refresh. everything else works fine. also fixed your remark with .get but didn't commit because of the current timeouts.
> 
> i guess it's because there are no explicit endpoints for form submissions. the contact property changes could be reworked if i use the endpoint for recently changed contacts. but the problem with the form submissions is that in order to get the values for a form submission one needs first to get every form available and afterwards get the form submission data. currently there is no specific endpoint to get every form submission independently from the form.

You can cache the response from forms? Or, you need to run a full refresh to forms?

tinomerl commented:
> > @marcosmarxm currently running in timeouts when running the acceptance tests for basic_read and full_refresh. everything else works fine. also fixed your remark with .get but didn't commit because of the current timeouts.
> > i guess it's because there are no explicit endpoints for form submissions. the contact property changes could be reworked if i use the endpoint for recently changed contacts. but the problem with the form submissions is that in order to get the values for a form submission one needs first to get every form available and afterwards get the form submission data. currently there is no specific endpoint to get every form submission independently from the form.
> 
> You can cache the response from forms? Or, you need to run a full refresh to forms?

@marcosmarxm how would i do the caching? I reduced the time range which will be called. last week i tried with about a year. also removed the `property_history` stream from the full_refresh since it creates a big output and is mostly the culprit for the timeouts. The only two tests failing right now are the read inputs in the acceptance tests. that is because we have a field in hubspot which contains a date but hubspot says it's property is a datetime. and the tests fail therefore. i committed everything and attached the logs.

<details>
<summary>Unit tests</summary>

```
python -m pytest unit_tests
Test session starts (platform: linux, Python 3.8.10, pytest 6.1.2, pytest-sugar 0.9.4)
cachedir: .pytest_cache
rootdir: /mnt/c/Users/Tino.Merl/Documents/repos/airbyte_dev, configfile: pytest.ini
plugins: sugar-0.9.4, timeout-1.4.2, requests-mock-1.8.0
collecting ... {"type": "LOG", "log": {"level": "INFO", "message": "Backing off get(...) for 0.0s (source_hubspot.errors.HubspotRateLimited: 429 Rate Limit Exceeded: API rate-limit has been reached until 0 seconds. See https://developers.hubspot.com/docs/api/usage-details)"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Rate limit reached. Sleeping for 0 seconds"}}

 airbyte-integrations/connectors/source-hubspot/unit_tests/test_client.py::test_client_backoff_on_limit_reached ✓                                                                                           2% ▍         {"type": "LOG", "log": {"level": "INFO", "message": "Backing off get(...) for 5.0s (requests.exceptions.HTTPError: 500 Server Error: None for url: https://api.hubapi.com/properties/v2/contact/properties?hapikey=wrong_key)"}}
{"type": "LOG", "log": {"level": "INFO", "message": "500 Server Error: None for url: https://api.hubapi.com/properties/v2/contact/properties?hapikey=wrong_key"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Caught retryable error after 1 tries. Waiting 5 more seconds then retrying..."}}

 airbyte-integrations/connectors/source-hubspot/unit_tests/test_client.py::test_client_backoff_on_server_error ✓                                                                                            5% ▌         {"type": "LOG", "log": {"level": "WARN", "message": "Stream `None` cannot be procced. This hapikey (THIS-IS-THE-API_KEY) does not have proper permissions! (requires any of [automation-access])"}}

 airbyte-integrations/connectors/source-hubspot/unit_tests/test_client.py::test_wrong_permissions_api_key ✓                                                                                                 8% ▊         
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_client.py::TestSplittingPropertiesFunctionality.test_splitting_properties ✓                                                                10% █         
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_client.py::TestSplittingPropertiesFunctionality.test_stream_with_splitting_properties ✓                                                    12% █▍        
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_client.py::TestSplittingPropertiesFunctionality.test_stream_with_splitting_properties_with_pagination ✓                                    15% █▌        
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_client.py::TestSplittingPropertiesFunctionality.test_stream_with_splitting_properties_with_new_record ✓                                    18% █▊        
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[string-expected0] ✓                                                            20% ██        
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[integer-expected1] ✓                                                           22% ██▍       
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[number-expected2] ✓                                                            25% ██▌       
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[bool-expected3] ✓                                                              28% ██▊       
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[boolean-expected4] ✓                                                           30% ███       
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[enumeration-expected5] ✓                                                       32% ███▍      
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[object-expected6] ✓                                                            35% ███▌      
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[array-expected7] ✓                                                             38% ███▊      
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[date-expected8] ✓                                                              40% ████      
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[date-time-expected9] ✓                                                         42% ████▍     
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[datetime-expected10] ✓                                                         45% ████▌     
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[json-expected11] ✓                                                             48% ████▊     
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_field_type_format_converting[phone_number-expected12] ✓                                                     50% █████     {"type": "LOG", "log": {"level": "WARN", "message": "Unsupported type _unsupported_field_type_ found"}}

 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_bad_field_type_converting[_unsupported_field_type_-expected0] ✓                                             52% █████▍    {"type": "LOG", "log": {"level": "WARN", "message": "Unsupported type None found"}}

 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_bad_field_type_converting[None-expected1] ✓                                                                 55% █████▌    {"type": "LOG", "log": {"level": "WARN", "message": "Unsupported type 1 found"}}

 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_bad_field_type_converting[1-expected2] ✓                                                                    57% █████▊    
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types0-some_field-None-None-None] ✓                                      60% ██████    
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types1-some_field-None-None-None] ✓                                      62% ██████▍   
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types2-some_field-None-None-None] ✓                                      65% ██████▌   
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types3-some_field-None-None-None] ✓                                      68% ██████▊   
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types4-some_field-None-None-None] ✓                                      70% ███████   
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[string-some_field-test-None-test] ✓                                                     72% ███████▍  
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types6-some_field-123.456-None-123.456] ✓                                75% ███████▌  
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types7-user_id-123-None-123] ✓                                           78% ███████▊  
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types8-some_field-123-None-123] ✓                                        80% ████████  
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types9-some_field--None-] ✓                                              82% ████████▍ 
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types10-some_field--None-None] ✓                                         85% ████████▌ 
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types11-some_field--None-None] ✓                                         88% ████████▊ 
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types12-some_field--None-None] ✓                                         90% █████████ 
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types13-some_field--None-None] ✓                                         92% █████████▍
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types14-some_field--date-time-None] ✓                                    95% █████████▌
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types15-some_field--date-time-] ✓                                        98% █████████▊
 airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_cast_type_if_needed[declared_field_types16-some_field-2020-date-time-2020] ✓                               100% ██████████
=================================================================================================== warnings summary ====================================================================================================
airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_bad_field_type_converting[_unsupported_field_type_-expected0]
airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_bad_field_type_converting[None-expected1]
airbyte-integrations/connectors/source-hubspot/unit_tests/test_field_type_converting.py::test_bad_field_type_converting[1-expected2]
  /mnt/c/Users/Tino.Merl/Documents/repos/airbyte_dev/airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:419: DeprecationWarning: The 'warn' method is deprecated, use 'warning' instead
    logger.warn(f"Unsupported type {field_type} found")

-- Docs: https://docs.pytest.org/en/stable/warnings.html

Results (8.87s):
      40 passed
```
</details>

<details>

<summary>Integration Tests</summary>

```
python -m pytest integration_tests
Test session starts (platform: linux, Python 3.8.10, pytest 6.1.2, pytest-sugar 0.9.4)
cachedir: .pytest_cache
rootdir: /mnt/c/Users/Tino.Merl/Documents/repos/airbyte_dev, configfile: pytest.ini
plugins: sugar-0.9.4, timeout-1.4.2, requests-mock-1.8.0
collecting ... 

Results (0.24s):
```

</details>

<details>

<summary>acceptance Tests</summary>

```
python -m pytest integration_tests -p integration_tests.acceptance
Test session starts (platform: linux, Python 3.8.10, pytest 6.1.2, pytest-sugar 0.9.4)
cachedir: .pytest_cache
rootdir: /mnt/c/Users/Tino.Merl/Documents/repos/airbyte_dev, configfile: pytest.ini
plugins: sugar-0.9.4, timeout-1.4.2, requests-mock-1.8.0
collecting ... {"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \nspec\ninput: /tmp/pytest-of-tino/pytest-15/test_match_expected_inputs0_0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_match_expected_inputs0_0/run_1/output"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_match_expected[inputs0] ✓                           6% ▋         
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_required[inputs0] ✓                                12% █▎        
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_optional[inputs0] ✓                                18% █▊        
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_has_secret[inputs0] ✓                              24% ██▍       
 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_secret_never_in_the_output[inputs0] ✓              29% ██▉       {"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \nspec\ninput: /tmp/pytest-of-tino/pytest-15/test_oauth_flow_parameters_inp0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_oauth_flow_parameters_inp0/run_1/output"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestSpec.test_oauth_flow_parameters[inputs0] ✓                   35% ███▌      {"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \ncheck --config tap_config.json\ninput: /tmp/pytest-of-tino/pytest-15/test_check_inputs0_0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_check_inputs0_0/run_1/output"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs0] ✓                             41% ████▎     {"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \ncheck --config tap_config.json\ninput: /tmp/pytest-of-tino/pytest-15/test_check_inputs1_0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_check_inputs1_0/run_1/output"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs1] ✓                             47% ████▊     {"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \ncheck --config tap_config.json\ninput: /tmp/pytest-of-tino/pytest-15/test_check_inputs2_0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_check_inputs2_0/run_1/output"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs2] ✓                             53% █████▍    {"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \ncheck --config tap_config.json\ninput: /tmp/pytest-of-tino/pytest-15/test_check_inputs3_0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_check_inputs3_0/run_1/output"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs3] ✓                             59% █████▉    {"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \ncheck --config tap_config.json\ninput: /tmp/pytest-of-tino/pytest-15/test_check_inputs4_0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_check_inputs4_0/run_1/output"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestConnection.test_check[inputs4] ✓                             65% ██████▌   {"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \ndiscover --config tap_config.json\ninput: /tmp/pytest-of-tino/pytest-15/test_discover_inputs0_0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_discover_inputs0_0/run_1/output"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_discover[inputs0] ✓                           71% ███████▏  {"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \ndiscover --config tap_config.json\ninput: /tmp/pytest-of-tino/pytest-15/test_defined_cursors_exist_in_0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_defined_cursors_exist_in_0/run_1/output"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestDiscovery.test_defined_cursors_exist_in_schema[inputs0] ✓    76% ███████▋  {"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \nread --config tap_config.json --catalog catalog.json\ninput: /tmp/pytest-of-tino/pytest-15/test_read_inputs0_0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_read_inputs0_0/run_1/output"}}
{"type": "LOG", "log": {"level": "ERROR", "message": "\nThe contacts stream has the following schema errors:\n2020-07-23 has invalid datetime format\n\nFailed validating 'format' in schema['properties']['properties']['properties']['p7_kontaktanfrage_datum']:\n    {'format': 'date-time', 'type': ['null', 'string']}\n\nOn instance['properties']['p7_kontaktanfrage_datum']:\n    '2020-07-23'"}}


――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― TestBasicRead.test_read[inputs0] ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

self = <source_acceptance_test.tests.test_core.TestBasicRead object at 0x7f6a37809d90>, connector_config = SecretDict(******)
configured_catalog = ConfiguredAirbyteCatalog(streams=[ConfiguredAirbyteStream(stream=AirbyteStream(name='campaigns', json_schema={'$schema...l_refresh'>, cursor_field=None, destination_sync_mode=<DestinationSyncMode.overwrite: 'overwrite'>, primary_key=None)])
inputs = BasicReadTestConfig(config_path='secrets/config.json', configured_catalog_path='sample_files/full_refresh_catalog.json', empty_streams={'workflows', 'property_history'}, expect_records=None, validate_schema=True, timeout_seconds=None)
expected_records = [], docker_runner = <source_acceptance_test.utils.connector_runner.ConnectorRunner object at 0x7f6a378015e0>
detailed_logger = <AirbyteNativeLogger detailed_logger acceptance_tests_logs/test_core.py__TestBasicRead__test_read[inputs0].txt (DEBUG)>

    def test_read(
        self,
        connector_config,
        configured_catalog,
        inputs: BasicReadTestConfig,
        expected_records: List[AirbyteMessage],
        docker_runner: ConnectorRunner,
        detailed_logger,
    ):
        output = docker_runner.call_read(connector_config, configured_catalog)
        records = [message.record for message in filter_output(output, Type.RECORD)]
    
        assert records, "At least one record should be read using provided catalog"
    
        if inputs.validate_schema:
>           self._validate_schema(records=records, configured_catalog=configured_catalog)

../../bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:262: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

records = [AirbyteRecordMessage(stream='campaigns', data={'id': 89263047, 'lastUpdatedTime': 1637571915271, 'appId': 2286, 'appN...red': 2, 'sent': 2, 'click': 2, 'open': 2}, 'type': 'AUTOMATED_EMAIL'}, emitted_at=1637587909000, namespace=None), ...]
configured_catalog = ConfiguredAirbyteCatalog(streams=[ConfiguredAirbyteStream(stream=AirbyteStream(name='campaigns', json_schema={'$schema...l_refresh'>, cursor_field=None, destination_sync_mode=<DestinationSyncMode.overwrite: 'overwrite'>, primary_key=None)])

    @staticmethod
    def _validate_schema(records: List[AirbyteRecordMessage], configured_catalog: ConfiguredAirbyteCatalog):
        """
        Check if data type and structure in records matches the one in json_schema of the stream in catalog
        """
        TestBasicRead._validate_records_structure(records, configured_catalog)
        bar = "-" * 80
        streams_errors = verify_records_schema(records, configured_catalog)
        for stream_name, errors in streams_errors.items():
            errors = map(str, errors.values())
            str_errors = f"\n{bar}\n".join(errors)
            logging.error(f"\nThe {stream_name} stream has the following schema errors:\n{str_errors}")
    
        if streams_errors:
>           pytest.fail(f"Please check your json_schema in selected streams {tuple(streams_errors.keys())}.")
E           Failed: Please check your json_schema in selected streams ('contacts',).

../../bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:207: Failed
----------------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------------
INFO     root:connector_runner.py:93 Docker run: 
read --config tap_config.json --catalog catalog.json
input: /tmp/pytest-of-tino/pytest-15/test_read_inputs0_0/run_1/input
output: /tmp/pytest-of-tino/pytest-15/test_read_inputs0_0/run_1/output
ERROR    root:test_core.py:204 
The contacts stream has the following schema errors:
2020-07-23 has invalid datetime format

Failed validating 'format' in schema['properties']['properties']['properties']['p7_kontaktanfrage_datum']:
    {'format': 'date-time', 'type': ['null', 'string']}

On instance['properties']['p7_kontaktanfrage_datum']:
    '2020-07-23'

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs0] ⨯                               82% ████████▎ {"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \nread --config tap_config.json --catalog catalog.json\ninput: /tmp/pytest-of-tino/pytest-15/test_read_inputs1_0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_read_inputs1_0/run_1/output"}}
{"type": "LOG", "log": {"level": "ERROR", "message": "\nThe contacts stream has the following schema errors:\n2020-07-23 has invalid datetime format\n\nFailed validating 'format' in schema['properties']['properties']['properties']['p7_kontaktanfrage_datum']:\n    {'format': 'date-time', 'type': ['null', 'string']}\n\nOn instance['properties']['p7_kontaktanfrage_datum']:\n    '2020-07-23'"}}


――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――― TestBasicRead.test_read[inputs1] ――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――――

self = <source_acceptance_test.tests.test_core.TestBasicRead object at 0x7f6a319a8880>, connector_config = SecretDict(******)
configured_catalog = ConfiguredAirbyteCatalog(streams=[ConfiguredAirbyteStream(stream=AirbyteStream(name='campaigns', json_schema={'$schema...l_refresh'>, cursor_field=None, destination_sync_mode=<DestinationSyncMode.overwrite: 'overwrite'>, primary_key=None)])
inputs = BasicReadTestConfig(config_path='secrets/config_oauth.json', configured_catalog_path='sample_files/configured_catalog_...treams={'campaigns', 'workflows', 'property_history'}, expect_records=None, validate_schema=True, timeout_seconds=None)
expected_records = [], docker_runner = <source_acceptance_test.utils.connector_runner.ConnectorRunner object at 0x7f6a3733c7c0>
detailed_logger = <AirbyteNativeLogger detailed_logger acceptance_tests_logs/test_core.py__TestBasicRead__test_read[inputs1].txt (DEBUG)>

    def test_read(
        self,
        connector_config,
        configured_catalog,
        inputs: BasicReadTestConfig,
        expected_records: List[AirbyteMessage],
        docker_runner: ConnectorRunner,
        detailed_logger,
    ):
        output = docker_runner.call_read(connector_config, configured_catalog)
        records = [message.record for message in filter_output(output, Type.RECORD)]
    
        assert records, "At least one record should be read using provided catalog"
    
        if inputs.validate_schema:
>           self._validate_schema(records=records, configured_catalog=configured_catalog)

../../bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:262: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

records = [AirbyteRecordMessage(stream='campaigns', data={'id': 89263047, 'lastUpdatedTime': 1637571915271, 'appId': 2286, 'appN...red': 2, 'sent': 2, 'click': 2, 'open': 2}, 'type': 'AUTOMATED_EMAIL'}, emitted_at=1637588004000, namespace=None), ...]
configured_catalog = ConfiguredAirbyteCatalog(streams=[ConfiguredAirbyteStream(stream=AirbyteStream(name='campaigns', json_schema={'$schema...l_refresh'>, cursor_field=None, destination_sync_mode=<DestinationSyncMode.overwrite: 'overwrite'>, primary_key=None)])

    @staticmethod
    def _validate_schema(records: List[AirbyteRecordMessage], configured_catalog: ConfiguredAirbyteCatalog):
        """
        Check if data type and structure in records matches the one in json_schema of the stream in catalog
        """
        TestBasicRead._validate_records_structure(records, configured_catalog)
        bar = "-" * 80
        streams_errors = verify_records_schema(records, configured_catalog)
        for stream_name, errors in streams_errors.items():
            errors = map(str, errors.values())
            str_errors = f"\n{bar}\n".join(errors)
            logging.error(f"\nThe {stream_name} stream has the following schema errors:\n{str_errors}")
    
        if streams_errors:
>           pytest.fail(f"Please check your json_schema in selected streams {tuple(streams_errors.keys())}.")
E           Failed: Please check your json_schema in selected streams ('contacts',).

../../bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:207: Failed
----------------------------------------------------------------------------- Captured log call -----------------------------------------------------------------------------
INFO     root:connector_runner.py:93 Docker run: 
read --config tap_config.json --catalog catalog.json
input: /tmp/pytest-of-tino/pytest-15/test_read_inputs1_0/run_1/input
output: /tmp/pytest-of-tino/pytest-15/test_read_inputs1_0/run_1/output
ERROR    root:test_core.py:204 
The contacts stream has the following schema errors:
2020-07-23 has invalid datetime format

Failed validating 'format' in schema['properties']['properties']['properties']['p7_kontaktanfrage_datum']:
    {'format': 'date-time', 'type': ['null', 'string']}

On instance['properties']['p7_kontaktanfrage_datum']:
    '2020-07-23'

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead.test_read[inputs1] ⨯                               88% ████████▉ {"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \nread --config tap_config.json --catalog catalog.json\ninput: /tmp/pytest-of-tino/pytest-15/test_sequential_reads_inputs0_0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_sequential_reads_inputs0_0/run_1/output"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \nread --config tap_config.json --catalog catalog.json\ninput: /tmp/pytest-of-tino/pytest-15/test_sequential_reads_inputs0_0/run_2/input\noutput: /tmp/pytest-of-tino/pytest-15/test_sequential_reads_inputs0_0/run_2/output"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_full_refresh.py::TestFullRefresh.test_sequential_reads[inputs0] ✓         94% █████████▌{"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \nread --config tap_config.json --catalog catalog.json\ninput: /tmp/pytest-of-tino/pytest-15/test_sequential_reads_inputs1_0/run_1/input\noutput: /tmp/pytest-of-tino/pytest-15/test_sequential_reads_inputs1_0/run_1/output"}}
{"type": "LOG", "log": {"level": "INFO", "message": "Docker run: \nread --config tap_config.json --catalog catalog.json\ninput: /tmp/pytest-of-tino/pytest-15/test_sequential_reads_inputs1_0/run_2/input\noutput: /tmp/pytest-of-tino/pytest-15/test_sequential_reads_inputs1_0/run_2/output"}}

 airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_full_refresh.py::TestFullRefresh.test_sequential_reads[inputs1] ✓        100% ██████████
========================================================================== short test summary info ==========================================================================
SKIPPED [1] ../../bases/source-acceptance-test/source_acceptance_test/plugin.py:56: Skipping TestIncremental.test_two_sequential_reads because not found in the config
FAILED ../../bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead::test_read[inputs0] - Failed: Please check your json_schema in selected...
FAILED ../../bases/source-acceptance-test/source_acceptance_test/tests/test_core.py::TestBasicRead::test_read[inputs1] - Failed: Please check your json_schema in selected...

Results (605.80s):
      15 passed
       2 failed
         - airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:247 TestBasicRead.test_read[inputs0]
         - airbyte-integrations/bases/source-acceptance-test/source_acceptance_test/tests/test_core.py:247 TestBasicRead.test_read[inputs1]
```

</details>

yevhenii-ldv commented:
Hi @tinomerl!
In order to increase the execution time of tests, and so that they are not interrupted by a timeout, you can specify the value of `timeout_seconds` in the `acceptance-test-config.yml` file. You can see how this is done for the `Prestashop` connector.

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/sample_files/configured_catalog_for_oauth_config.json:
```diff
@@ -75,6 +75,15 @@
       "destination_sync_mode": "overwrite"
     },
+    {
+      "stream": {
```
Please, add `property_history` stream too

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/sample_files/full_refresh_catalog.json:
```diff
@@ -72,6 +72,15 @@
       "sync_mode": "full_refresh",
       "destination_sync_mode": "overwrite"
     },
+    {
```
Please, add `property_history` stream too

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:
```diff
@@ -316,7 +316,7 @@ def _field_to_datetime(value: Union[int, str]) -> pendulum.datetime:
         """Skip records that was updated before our start_date"""
         for record in records:
-            updated_at = record[self.updated_at_field]
+            updated_at = record.get(self.updated_at_field)
```
Why do we specify such a safe data retrieval method, we may receive a record that does not contain the `updated_at_field` field

tinomerl replied:
Reverted it. I changed because i got an issue calling data from our hubspot instance. Therefore i changed it to .get. But i reran it with a smaller timeframe with the reverted change. No issues. So it seems The issue was in our hubspot instance.

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:
```diff
@@ -699,6 +699,76 @@ class FormStream(Stream):
+    Docs: https://legacydocs.hubspot.com/docs/methods/contacts/get_contacts
+    """
+    more_key = "has-more"
+    url = "/contacts/v1/lists/all/contacts/all"
```
We do not use the `entity` variable to get all properties for this stream, this stream cannot receive additional `properties` by passing them as parameters?

tinomerl replied:
See answer [here](https://github.com/airbytehq/airbyte/pull/7787#discussion_r782173372)

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:
```diff
@@ -699,6 +699,76 @@ class FormStream(Stream):
+    Get the Submissions for a form and the field values.
+    Docs: https://legacydocs.hubspot.com/docs/methods/forms/get-submissions-for-a-form
+    """
+    url = "/form-integrations/v1/submissions/forms"
```
Same question as [here](https://github.com/airbytehq/airbyte/pull/7787/files#r765321373)

tinomerl replied:
This stream has merely two real columns which hold the most information after it's transformed. One for a name of a form field and one for the value which was submitted. Otherwise we would have a huge table with n columns where n is the number of different form fields accross all forms. Also with every new form the schema would need to be updated. 
This goes also for the contact PropertyHistory

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:
```diff
@@ -699,6 +699,76 @@ class FormStream(Stream):
+    """
+    url = "/form-integrations/v1/submissions/forms"
+    limit = 50
+    updated_at_field = "submittedAt"
```
if we set updated_at_field, why we cannot inherite this class from `IncrementalStream`, not from `Stream`? Can't this stream be incremental?

tinomerl replied:
The form Submission stream can't be incremental since there is no real endpoint for all the form submissions. you need to call the endpoint per form and the endpoint doesn't support filtering. Therefore you always have to call every form and all of their submissions.

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:
```diff
@@ -699,6 +699,76 @@ class FormStream(Stream):
+                        yield version
+
+
+class FormSubmssionStream(Stream):
```
```suggestion
class FormSubmissionStream(Stream):
```

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:
```diff
@@ -699,6 +699,76 @@ class FormStream(Stream):
+
+    def _transform(self, records: Iterable) -> Iterable:
+        for record in records:
+            values: dict = record.get("values") or None
+            if values:
+                record["submittedAt"] = self._field_to_datetime(record["submittedAt"]).to_datetime_string()
```
```suggestion
            if record.get("values"):
                record["submittedAt"] = self._field_to_datetime(record["submittedAt"]).to_datetime_string()
```

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:
```diff
@@ -699,6 +699,76 @@ class FormStream(Stream):
+    page_filter = "vidOffset"
+
+    def list(self, fields) -> Iterable:
+        properties = self._api.get(f"/properties/v2/contact/properties")
```
I don’t understand why we can’t just define the `contact` for this stream as `self.entity` and not override the `list` method, it will be executed as in other streams, right?

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:
```diff
@@ -699,6 +699,76 @@ class FormStream(Stream):
+        yield from self.read(partial(self._api.get, url=self.url), params)
+
+    def _transform(self, records: Iterable) -> Iterable:
+        record: Dict[str, Dict]
```
```suggestion
        
```
I think that this line does not carry any information we need, it is rather superfluous than useful

tinomerl replied:
Removed

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:
```diff
@@ -699,6 +699,76 @@ class FormStream(Stream):
+    page_filter = "after"
+    
+    def list(self, fields) -> Iterable:
+        params = {
+            "limit": 50,
+        }
+        forms = self.read(partial(self._api.get, url="/marketing/v3/forms"), params)
+        for row in forms:
+            record = self.read(partial(self._api.get, url=f"{self.url}/{row['id']}"), params)
```
```suggestion
        forms = self.read(partial(self._api.get, url="/marketing/v3/forms"))
        for row in forms:
            record = self.read(partial(self._api.get, url=f"{self.url}/{row['id']}"))
```
There is no need to explicitly pass only the `limit` in the parameters, since we will get it from the class variable anyway

tinomerl replied:
Removed

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:
```diff
@@ -699,6 +699,76 @@ class FormStream(Stream):
+        params = {
+            "limit": 50,
+        }
+        forms = self.read(partial(self._api.get, url="/marketing/v3/forms"), params)
+        for row in forms:
+            record = self.read(partial(self._api.get, url=f"{self.url}/{row['id']}"), params)
```
```suggestion
        forms_stream = FormStream(api=self._api, start_date=str(self._start_date))
        for row in forms_stream.list(fields={}):
```
I think, it will be more clearly, WDYT?

tinomerl replied:
is better. i added it.

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/api.py:
```diff
@@ -699,6 +699,76 @@ class FormStream(Stream):
+            value_dict: Dict
+            for key, value_dict in properties.items():
+                versions = value_dict.get("versions")
+                if key == "lastmodifieddate":
+                    continue
```
Could you please leave a comment as to why we are skipping processing for the `lastmodifieddate` key?

tinomerl replied:
comment is added.

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/schemas/form_submissions.json:
```diff
@@ -0,0 +1,35 @@
+            },
+            "objectTypeId": {
+              "value": {
+                "type": ["null", "string"]
```
Are u sure, that field which consists `Id` in name has `string` type?

tinomerl replied:
Yeah HubSpot returns a value which consists only of this string "0-1". Therefore String.

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/schemas/form_submissions.json:
```diff
@@ -0,0 +1,35 @@
+        "type": ["null", "string"]
+      }
+    }
+  }
```
please, add empty string in the end of file

tinomerl replied:
added

yevhenii-ldv commented on airbyte-integrations/connectors/source-hubspot/source_hubspot/schemas/property_history.json:
```diff
@@ -0,0 +1,35 @@
+      "source-type": {
+        "type": ["null", "string"]
+      },
+      "source-id": {
```
same question as [here](https://github.com/airbytehq/airbyte/pull/7787/files#r765338260)

tinomerl replied:
A Property can either be changed by workflow, form or manully. The id for either is saved in the `source-id` . The thin is, that for workflow the id is really a id as in integer. but the id of a form is a mixture of characters and numbers. Therefore it is safer to just cast this as a string.

yevhenii-ldv reviewed (changes requested):
See my comments.
Also, please, run command `./gradlew format` to format project. This will format the changes in the project to the required template.
_**NOTE:**_ If the formatting process also affects files in which you did not make changes, please roll back the changes to those files to the state in which they were before. Thanks 😃

tinomerl commented:
Hey @yevhenii-ldv thanks for the Feedback! I'll get to working to after the 19th since i'm not available next week. Just wanted to give you a heads up and let u know. :)

marcosmarxm commented:
@tinomerl I added the label `blocked` let me know when you return working on this.

tinomerl commented:
Hey @yevhenii-ldv 

sorry for taking so long to work on your feedback. i implemented everything. The tests still throw the same error for contact lists mentioned [here](https://github.com/airbytehq/airbyte/pull/7787#issuecomment-975543854). But additionally i get a sequential read error for the property history stream. The problem lies in the calculated fields like `hs_time_in_subscriber`. I tried filtering it out via the `acceptance-test-config.yml`. The thing is the field which is calculated is a value and not a field in the final table. so it looks like this :

|property|value|changed_date|
|---|---|---|
|hs_time_in_subscriber|123456789|2021-01-01T00:00:00Z

The second call will of course have a higher value for the same property. Is there a way to filter out a field if it contains a specific value?

marcosmarxm commented:
@tinomerl can you solve the conflicts? Other PR add the form submissions stream.

tinomerl commented:
@marcosmarxm solved the conflicts
